File: client/agentsec-cli/internal/auth/token.go
package auth

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type TokenInfo struct {
	Raw       string
	Claims    map[string]any
	Valid     bool
	Status    string
	ExpiresAt *time.Time
}

func cachePath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".agentsec", "agent.token"), nil
}

func ReadToken() (string, error) {
	if token := strings.TrimSpace(os.Getenv("AGENTSEC_TOKEN")); token != "" {
		return token, nil
	}
	path, err := cachePath()
	if err != nil {
		return "", err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(data)), nil
}

func WriteToken(token string) error {
	path, err := cachePath()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(strings.TrimSpace(token)), 0o600)
}

func InspectToken(token string) TokenInfo {
	info := TokenInfo{
		Raw:    strings.TrimSpace(token),
		Claims: map[string]any{},
		Status: "missing",
	}
	if info.Raw == "" {
		return info
	}
	parts := strings.Split(info.Raw, ".")
	if len(parts) < 2 {
		info.Status = "invalid"
		return info
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		info.Status = "invalid"
		return info
	}
	if err := json.Unmarshal(payload, &info.Claims); err != nil {
		info.Status = "invalid"
		return info
	}
	info.Valid = true
	info.Status = "valid"
	if expValue, ok := info.Claims["exp"]; ok {
		switch value := expValue.(type) {
		case float64:
			ts := time.Unix(int64(value), 0).UTC()
			info.ExpiresAt = &ts
		case json.Number:
			if n, err := value.Int64(); err == nil {
				ts := time.Unix(n, 0).UTC()
				info.ExpiresAt = &ts
			}
		}
	}
	if info.ExpiresAt != nil && !info.ExpiresAt.After(time.Now().UTC()) {
		info.Status = "expired"
	}
	return info
}

func MustReadAndInspect() (TokenInfo, error) {
	token, err := ReadToken()
	if err != nil {
		return TokenInfo{}, err
	}
	info := InspectToken(token)
	if !info.Valid && info.Status != "expired" {
		return info, fmt.Errorf("invalid token format")
	}
	return info, nil
}

File: client/agentsec-python-sdk/agentsec/auth/token.py
from __future__ import annotations

import base64
import json
import os
import time
from pathlib import Path
from typing import Any, Optional


class TokenManager:
    CACHE_PATH = Path.home() / ".agentsec" / "agent.token"

    def load_token(self) -> Optional[str]:
        token = os.getenv("AGENTSEC_TOKEN")
        if token:
            return token.strip()
        return self.read_cached_token()

    def read_cached_token(self) -> Optional[str]:
        try:
            if self.CACHE_PATH.exists():
                token = self.CACHE_PATH.read_text(encoding="utf-8").strip()
                return token or None
        except OSError:
            return None
        return None

    def cache_token(self, token: str) -> None:
        self.CACHE_PATH.parent.mkdir(parents=True, exist_ok=True)
        self.CACHE_PATH.write_text(token.strip(), encoding="utf-8")

    def clear_cached_token(self) -> None:
        try:
            self.CACHE_PATH.unlink(missing_ok=True)
        except OSError:
            return

    def parse_claims(self, token: str) -> dict[str, Any]:
        parts = token.split(".")
        if len(parts) < 2:
            return {}
        payload = parts[1] + "=" * (-len(parts[1]) % 4)
        try:
            decoded = base64.urlsafe_b64decode(payload.encode("utf-8"))
            return json.loads(decoded.decode("utf-8"))
        except Exception:
            return {}

    def token_status(self, token: Optional[str]) -> str:
        if not token:
            return "missing"
        claims = self.parse_claims(token)
        if not claims:
            return "invalid"
        exp = claims.get("exp")
        if exp is None:
            return "valid"
        try:
            if float(exp) <= time.time():
                return "expired"
        except (TypeError, ValueError):
            return "invalid"
        return "valid"
